ffcx/codegeneration/integrals.py

```diff
@@ -691,3 +688,29 @@ def generate_block_parts(self, quadrature_rule: QuadratureRule, blockmap: Tuple,
         quadparts += [pre_loop, hoist, body]
 
         return preparts, quadparts
+
+    def fuse_loops(self, definitions):
+        """
+        Merge a sequence of loops with the same iteration space into a single loop.
+        Loop fusion improves data locality, cache reuse and decreases the loop control overhead.
+        NOTE: Loop fusion might increase the pressure on register allocation.
+        Ideally, we should define a cost function to determine how many loops should fuse at a time.
+        """
+        L = self.backend.language
+
+        loops = collections.defaultdict(list)
+        pre_loop = []
+        for access, definition in definitions.items():
+            for d in definition:
+                if isinstance(d, L.ForRange):
+                    loops[(d.begin, d.end)] += [d.body]
```

Comment:
I think you should add ForRange's iteration symbol (e.g. "ic") to dicts keys. The function looks general and could accidentally fuse across different loops if used elsewhere.

Reply:
Added the index to the dict key:
```python 
loops[(d.index, d.begin, d.end)] += [d.body]
```